# Tutorial: Real Numbers and Their Operations in Python

Python makes working with real numbers easy and efficient. This tutorial will guide you through the concepts of integers, rational and irrational numbers, basic arithmetic operations, and fundamental properties of real numbers—all using Python.

---

## **1. Understanding Real Numbers**
Real numbers include:

- **Integers** (whole numbers): `-3, 0, 5`
- **Rational numbers** (fractions/decimals that can be written as a ratio of two integers): `1/2, 3.75`
- **Irrational numbers** (cannot be written as a ratio of two integers): `π, √2`

Python supports all these number types with different representations:

In [1]:
# Integer example
a = 5

# Rational number (float)
b = 3.75

# Irrational number approximation (math module)
import math
c = math.sqrt(2)
print(c)  # Approximation of √2

1.4142135623730951


## **2. Basic Arithmetic Operations in Python**
Python provides simple operators for arithmetic operations:

In [11]:
# Addition
sum_result = 5 + 3

# Subtraction
diff_result = 10 - 4

# Multiplication
product_result = 6 * 7

# Division (float result)
quotient_result = 37 / 4

print('5+3=' + str(sum_result), '   10-4='+str(diff_result), '   6*7='+str(product_result), '   15/4='+str(quotient_result))

5+3=8    10-4=6    6*7=42    15/4=9.25


You can also perform **integer** division using // and **modulus** using %:

In [15]:
# Integer division (floor division)
integer_div = 37 // 4  # Result: 9

# Modulus (remainder)
remainder = 37 % 4  # Result: 1

print('37 // 4=' + str(integer_div), '     37 % 4=' + str(remainder))

37 // 4=9      37 % 4=1


### Integer division
* **1. Division Step:** Normally, dividing 37 by 4 gives 9.25.

* **2. Flooring Step:** Since we're using floor division (//), the decimal part is discarded, leaving only the whole number.

* **3. Final Result:** integer_div is assigned the value 9.

Floor division is useful when you need to work with whole numbers, like calculating how many full boxes can be packed with a given number of items or determining how many times something fits into a given space.

### Modulus (remainder)

* **1. Division Step:** Normally, 37 ÷ 4 equals 9.25.

* **2. Multiplication Step:** The integer quotient is 9, so multiplying 9 × 4 = 36.

* **3. Finding the Remainder:** Subtract 36 from 37 to get 1.

So, remainder is assigned the value 1.

## **Real numbers follow some basic properties:**

### **Distributive Property**

For any numbers a, b, and c:


 $ a \times (b + c) = (a \times b) + (a \times c ) $

In [17]:
a, b, c = 2, 3, 4
left_side = a * (b + c)
right_side = (a * b) + (a * c)

print(left_side == right_side)  # True

True


### **Associative Property**
* **Addition:** $ (a + b) + c = a + (b + c) $           
         
* **Multiplication:** $ (a \times b) \times c = a \times (b \times c) $

In [19]:
a, b, c = 1, 2, 3

# Addition associative
addition_assoc = (a + b) + c == a + (b + c)

# Multiplication associative
multiplication_assoc = (a * b) * c == a * (b * c)

print(addition_assoc, multiplication_assoc)  # True, True

True True


### **Commutative Property**
* **Addition:** $ a + b = b + a $        

* **Multiplication:** $ a \times b = b \times a $

In [20]:
a, b = 5, 10

# Addition commutative
addition_comm = (a + b) == (b + a)

# Multiplication commutative
multiplication_comm = (a * b) == (b * a)

print(addition_comm, multiplication_comm)  # True, True

True True


### **Identity Property**
* **Additive Identity:** $ a + 0 = a $      

* **Multiplicative Identity:** $ a \times 1 = a $

In [21]:
a = 5

# Addition Identity
addition_identity = (a + 0) == a

# Multiplicative Identity
multiplicative_identity = (a * 1) == a

print(addition_identity, multiplicative_identity) # True, True

True True


### 5. **Inverse Property**
* **Additive Inverse:** $ a + (-a) = 0 $    

* **Multiplicative Inverse:** $ a \times \frac{1}{a} = 1 $    (for $ a \neq 0 $)

In [23]:
def additive_inverse(a):
    return -a

def multiplicative_inverse(a):
    if a == 0:
        return "Undefined (division by zero is not allowed)"
    return 1 / a

# Example usage:
num = 5

print(f"Additive Inverse of {num}: {additive_inverse(num)}")
print(f"Multiplicative Inverse of {num}: {multiplicative_inverse(num)}")


Additive Inverse of 5: -5
Multiplicative Inverse of 5: 0.2


### **Closure Property**
* **Addition:** The sum of any two real numbers is a real number.
* **Multiplication:** The product of any two real numbers is a real number.

The Closure Property states that a set is closed under an operation if performing that operation on members of the set always results in a member of the same set. For example:

The set of integers (
𝑍
) is closed under addition because the sum of any two integers is always an integer.

The set of real numbers (
𝑅
) is closed under multiplication because the product of any two real numbers is always a real number.

In [24]:
def is_closed_under_addition(numbers):
    for i in numbers:
        for j in numbers:
            if i + j not in numbers:
                return False
    return True

def is_closed_under_multiplication(numbers):
    for i in numbers:
        for j in numbers:
            if i * j not in numbers:
                return False
    return True

# Example usage:
num_set = {0, 1, 2, 3, 4}

print(f"Set {num_set} is closed under addition: {is_closed_under_addition(num_set)}")
print(f"Set {num_set} is closed under multiplication: {is_closed_under_multiplication(num_set)}")


Set {0, 1, 2, 3, 4} is closed under addition: False
Set {0, 1, 2, 3, 4} is closed under multiplication: False


## [College Algebra Context](./README.md)